### PR TITLE
Book clubs update and search features

### PR DESCRIPTION
**Pull request by ChrisThackrey**
-- Finished Search Bar Functionality
-- Linked main bookclub view with details view
-- Added actions and reducer to AppContext
-- Implemented Fuzzy Search Enhancement
-- other small stuff

### CODE REVIEW

**Review by mattBman23 — APPROVED** [review]

File: client/components/bookClubPage/BookClubItem.jsx [hunk 1]
```diff
@@ -1,29 +1,25 @@
...
+  const { club, getClubById } = useContext(AppContext);
+
+  const handleClicked = () => getClubById(current._id);
 
   return (
-    <Link to={`/clubs/${id}`}>
-      <Container className='my-1 p-1'>
-        {/* <Card.ImgOverlay>
+    <Container className='my-1 p-1' onClick={handleClicked}>
+      {/* <Card.ImgOverlay>
```
Comment (jfleming9357) [inline client/components/bookClubPage/BookClubItem.jsx:12]: Should the click event be on the container or the image? When I click anywhere on the carousel I get 404 errors for trying to get /bookclubs/undefined

File: client/context/context.jsx [hunk 2]
```diff
@@ -1,11 +1,240 @@
...
+  async function leaveClubById(id, userId) {
+    try {
+      await axios.post(`/bookclub/join/${id}`, {
+        userId
+      });
+
+      dispatch({
+        type: 'LEAVE_BOOKCLUB',
+        payload: user._id
+      });
```
Comment (jfleming9357) [inline client/context/context.jsx:169]: Oops wrong spot throwing the error. The above two functions both go to the catch block, and then throw an error in that catch block of trying to read 'data' of undefined

**Review by jfleming9357 — CHANGES_REQUESTED** [review]
Some bugs that need to be fixed. I don't love the fact that state is getting copied and reset to update one field of state every time in the context file, but I don't know enough about the method you are using to say whether there is a better way to do it. Would it be more efficient to have a lot of useState hooks? There would be quite a lot but at least you wouldn't have to copy the entire state object. Again don't know what the best practice for this is, just a shot in the dark.

**Comment by ChrisThackrey** [comment]
> Some bugs that need to be fixed. I don't love the fact that state is getting copied and reset to update one field of state every time in the context file, but I don't know enough about the method you are using to say whether there is a better way to do it. Would it be more efficient to have a lot of useState hooks? There would be quite a lot but at least you wouldn't have to copy the entire state object. Again don't know what the best practice for this is, just a shot in the dark.

Refactored to a bunch of useState's, removed useReducer, still works and doesn't copy entire state object.

File: client/context/context.jsx (hunk 2 above)
Comment (ChrisThackrey) [reply client/context/context.jsx]: Which one was this? Not using dispatch/reducer now. This one may not using the global state setup on Daniel's end, I think he was on join/leave functionality.

File: client/components/bookClubPage/BookClubItem.jsx (hunk 1 above)
Comment (ChrisThackrey) [reply client/components/bookClubPage/BookClubItem.jsx:15]: change request handled

**Review by jfleming9357 — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**